src/charm.py

```diff
@@ -381,6 +387,35 @@ def _on_update_status(self, _) -> None:
             # Set active status when primary is known
             self.app.status = ActiveStatus()
 
+    def _on_cos_agent_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the cos_agent relation created event.
+
+        Enable the mysqld-exporter snap service.
+        """
+        if not self._is_peer_data_set:
+            event.defer()
+            return
+
+        try:
+            self._mysql.connect_mysql_exporter()
+        except MySQLExporterConnectError:
+            logger.exception("Unable to set up the mysqld-exporter snap service")
+            self.unit.status = BlockedStatus("Unable to start mysqld-exporter")
+
+    def _on_cos_agent_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle the cos_agent relation broken event.
+
+        Disable the mysqld-exporter snap service.
+        """
+        if not self._is_peer_data_set:
+            return
+
+        try:
+            self._mysql.stop_mysql_exporter()
+        except MySQLExporterConnectError:
+            logger.exception("Unable to stop the mysqld-exporter snap service")
+            self.unit.status = BlockedStatus("Unable to stop mysqld-exporter")
+
```

Comment:
IMHO.
1) add debug log when you are deferring an event (line 396). It is very hard to guess in production why things didn't happen.
2) IMHO, we should rise Error and not Blocked if we failed to start/stop. User will be able to call "juju resolve", otherwise charm will stuck in blocked state if it failed to stop mysqld-exporter. Isn't it?

Reply:
I 2nd point 2, specially because the exporter is a not the main charm service

Reply:
addressed in 9f7d5c199f5e6824c0f44babf3c7917c58930592